src/model/encoding/convertFromRawToDraftState.js

```diff
@@ -13,60 +13,229 @@
+          Array.isArray(block.children),
+          'invalid RawDraftContentBlock can not be converted to ContentBlockNode',
+        );
+
+        // ensure children have valid keys to enable sibling links
+        const children = block.children.map(addKeyIfMissing);
+
+        // root level nodes
+        const contentBlockNode = new ContentBlockNode({
+          ...decodeBlockNodeConfig(block, entityMap),
+          prevSibling: index === 0 ? null : blocks[index - 1].key,
+          nextSibling:
+            index === blocks.length - 1 ? null : blocks[index + 1].key,
+          children: List(children.map(child => child.key)),
+        });
+
+        // push root node to blockMap
+        acc.push(contentBlockNode);
+
+        // this stack is used to ensure we visit all nodes respecting depth ordering
+        let stack = updateNodeStack([], children, contentBlockNode);
+
+        // start computing children nodes
+        while (stack.length > 0) {
+          // we pop from the stack and start processing this node
+          const node = stack.pop();
+
+          // parentRef already points to a converted ContentBlockNode
+          const parentRef: ContentBlockNode = node.parentRef;
+          const siblings = parentRef.getChildKeys();
+          const index = siblings.indexOf(node.key);
+          const isValidBlock = Array.isArray(node.children);
 
-    var inlineStyles = decodeInlineStyleRanges(text, inlineStyleRanges);
+          if (!isValidBlock) {
+            invariant(
+              isValidBlock,
+              'invalid RawDraftContentBlock can not be converted to ContentBlockNode',
+            );
+            break;
+          }
 
-    // Translate entity range keys to the DraftEntity map.
-    var filteredEntityRanges = entityRanges
-      .filter(range => fromStorageToLocal.hasOwnProperty(range.key))
-      .map(range => {
-        return {...range, key: fromStorageToLocal[range.key]};
-      });
+          // ensure children have valid keys to enable sibling links
+          const children = node.children.map(addKeyIfMissing);
 
-    var entities = decodeEntityRanges(text, filteredEntityRanges);
-    var characterList = createCharacterList(inlineStyles, entities);
+          const contentBlockNode = new ContentBlockNode({
+            ...decodeBlockNodeConfig(node, entityMap),
+            parent: parentRef.getKey(),
+            children: List(children.map(child => child.key)),
+            prevSibling: index === 0 ? null : siblings.get(index - 1),
+            nextSibling:
+              index === siblings.size - 1 ? null : siblings.get(index + 1),
+          });
 
-    return new ContentBlock({key, type, text, depth, characterList, data});
+          // push node to blockMap
+          acc.push(contentBlockNode);
+
+          // this stack is used to ensure we visit all nodes respecting depth ordering
+          stack = updateNodeStack(stack, children, contentBlockNode);
+        }
+
+        return acc;
+      }, [])
```

Comment:
Nice catch, actually the return type should be array of blocks since we use this later on:

 ContentState.createFromBlockArray

Reply:
The reason why flow does not fail is due to the above method been union of Array and BlockMap

Reply:
Actually it only accepts Array, so we should just need to update the return type of this

‘’’javascript

var BlockMapBuilder = {
  createFromArray: function(blocks: Array<BlockNodeRecord>): BlockMap {
    return OrderedMap(blocks.map(block => [block.getKey(), block]));
  },
};
‘’’

Reply:
Actually all that is doing is wrapping it on a immutable OrderedMap, will update this Pr to do that in place here instead of returning the Array and having to do that again

Sorry for the many comments in here, hard to reply now since been using my phone and don’t have a laptop near me atm

Reply:
Actually all that is doing is wrapping it on a immutable OrderedMap, will update this Pr to do that in place here instead of returning the Array and having to do that again

Sorry for the many comments in here, hard to reply now since been using my phone and don’t have a laptop near me atm